api/sessionrecording/gzip_reader.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2024 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sessionrecording
+
+import (
+	"compress/gzip"
+	"io"
+
+	"github.com/gravitational/trace"
+)
+
+// gzipReader wraps file, on close close both gzip writer and file
+type gzipReader struct {
+	io.ReadCloser
+	inner io.ReadCloser
+}
+
+// Close closes file and gzip writer
+func (f *gzipReader) Close() error {
+	var errors []error
+	if f.ReadCloser != nil {
+		errors = append(errors, f.ReadCloser.Close())
+		f.ReadCloser = nil
+	}
+	if f.inner != nil {
+		errors = append(errors, f.inner.Close())
+		f.inner = nil
+	}
+	return trace.NewAggregate(errors...)
+}
+
+func newGzipReader(reader io.ReadCloser) (*gzipReader, error) {
```

Comment:
On closer look, we made our own `gzipReader` instead of just using Go's `gzip.Reader` directly because we wanted to accept a `ReadCloser` and close it.

I wonder what it would look like if we shed our custom `gzipReader` completely and just had the `sessionrecording.Reader` keep a reference to the underlying reader to close it.

(Or even do what the gzip package does, and accept an `io.Reader` instead of an `io.ReadCloser`, and leave it up to the client to close things down if necessary)

Reply:
Really great idea. Thank you for the thorough review!

I started with changing our `gzipReader` to only accept an `io.Reader` and realized sessionrecording was always using `io.NopCloser`, so that was an easy change. And then it was mostly straight forward to remove `gzipReader` entirely and now `sessionrecording.Reader` constructs a `gzip.Reader` with multistream disabled. I need to come back to this because I do not like the name `inner`.

Reviewing by commit instead of the whole diff might make the above easier to see.